File: ch4/4-10.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <math.h>
#include <string.h>

#define MAXVAL 100
#define BUFSIZE 100
#define MAXLINE 1000  // length of a maximum line to be read
#define MAXOP 100     // max size of operand or operator
#define NUMBER '0'    // signal that a number wes found
#define PRINT 'p'     // print the element witout poping
#define DUPLICATE 'd' // duplicate element
#define SWAP 's'      // swap the top two elements
#define CLEAR 'c'     // clear the stack
#define VARIABLE 'x'  // name of the variable for the latest printed value
#define NAME 'n'      // name of the mathematical operation

int sp = 0;
double val[MAXVAL];
double x;

char buf[BUFSIZE];
int bufp = 0;

int get_line(char s[], int len);
int getop(char s[], char line[]);
void push(double f);
double pop(void);
void peek(void);
void dup(void);
void swap(void);
void clear(void);
void mathfunc(char s[]);

int main(void)
{
    int type, operator;
    double op2;
    char s[MAXOP], line[MAXLINE];

    while (get_line(line, MAXLINE) > 0) {
        while ((type = getop(s, line)) != '\0') {
            switch (type) {
                case NUMBER:
                    operator = 1;
                    push(atof(s));
                    break;
                case NAME:
                    operator = 1;
                    mathfunc(s);
                    break;
                case VARIABLE:
                    operator = 1;
                    push(x);
                    break;
                case '+':
                    operator = 1;
                    push(pop() + pop());
                    break;
                case '*':
                    operator = 1;
                    push(pop() * pop());
                    break;
                case '-':
                    operator = 1;
                    op2 = pop();
                    push(pop() - op2);
                    break;
                case '/':
                    operator = 1;
                    op2 = pop();
                    if (op2 != 0.0) {
                        push(pop() / op2);
                    } else {
                        printf("error: zero divisor\n");
                    }
                    break;
                case '%':
                    operator = 1;
                    op2 = pop();
                    if (op2 != 0) {
                        push(fmod(pop(), op2));
                    }
                    break;
                case PRINT:
                    operator = 0;
                    peek();
                    break;
                case DUPLICATE:
                    operator = 0;
                    dup();
                    break;
                case SWAP:
                    operator = 0;
                    swap();
                    break;
                case CLEAR:
                    operator = 0;
                    clear();
                    break;
                case '\n':
                    if (operator) {
                      x = pop();
                      printf("\t%.8g\n", x);
                    }
                    break;
                default:
                    printf("error: unknown command %s\n", s);
            }
        }
    }

    return 0;
}

void mathfunc(char s[])
{
    double op2;
    if (strcmp(s, "sin") == 0) {
        push(sin(pop()));
    } else if (strcmp(s, "cos") == 0) {
        push(cos(pop()));
    } else if (strcmp(s, "exp") == 0) {
        push(exp(pop()));
    } else if (strcmp(s, "pow") == 0) {
        op2 = pop();
        push(pow(pop(), op2));
    } else {
        printf("error: %s not supported\n", s);
    }
}

/* peek: prints without poping */
void peek(void)
{
    x = val[sp-1];
    printf("%g\n", x);
}

/* duplicate: duplicates the last element */
void dup(void)
{
    push(val[sp]);
}

/* swap: swaps the top 2 elements */
void swap()
{
    double temp = val[sp-2];
    val[sp-2] = val[sp-1];
    val[sp-1] = temp;
}

/* clear: clears the stack */
void clear(void)
{
    sp = 0;
}

int get_line(char s[], int len)
{
    int c, i;
    i = 0;
    while(--len > 0 && ((c = getchar()) != EOF && c != '\n')) {
        s[i++] = c;
    }
    if (c == '\n') {
        s[i++] = '\n';
    }
    s[i] = '\0';

    return i;
}

int j = 0;
int getop(char s[], char line[])
{
    int i, c;
    if (line[j] == '\n') {
        j++;
        return '\n';
    }
    while (((s[0] = c = line[j++]) != '\0') && (((c == ' ') || (c == '\t')))) {
        //
    }
    s[1] = '\0';
    i = 0;
    if (islower(c)) {
        while (islower(s[++i] = c = line[j++])) {
            //
        }
        s[i] = '\0';
        if (strlen(s) > 1) {
            return NAME;
        }
        return c;
    }
    if (c == '\0') {
        j = 0;
        return c;
    }
    if (!isdigit(c) && c != '.') {
        return c;
    }
    if (isdigit(c)) {
        while (isdigit(s[++i] = c = line[j++])) {
            //
        }
    }
    if (c == '.') {
        while (isdigit(s[++i] = c = line[j++])) {

        }
    }
    s[i] = '\0';

    return NUMBER;
}

void push(double f)
{
    if (sp < MAXVAL) {
        val[sp++] = f;
    } else {
        printf("error: stack full, can't push %g\n", f);
    }
}

double pop(void)
{
    if (sp > 0) {
        return val[--sp];
    } else {
        printf("error: stack empty\n");
        return 0.0;
    }
}

File: ch4/4-10.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"unicode"
)

const (
	MaxVal  = 100
	BufSize = 100
	MaxOp   = 100
	MaxLine = 1000
)

var sp, bufp int
var buf []byte
var val []float64
var x float64

func main() {
	var operator bool
	var v rune
	buf = make([]byte, BufSize)
	val = make([]float64, MaxVal)
	s := make([]byte, 0, MaxOp)
	vars := make([]float64, MaxVal)

	for i := 0; i < 26; i++ {
		vars[i] = 0.0
	}

	for {
		line, l := getLine(MaxLine)
		if l <= 0 {
			break
		}
	GetOp:
		for {
			typ, err := getop(&s, line)
			if err != nil {
				break
			}
			switch typ {
			case 'n':
				n, err := strconv.ParseFloat(string(s), 64)
				if err != nil {
					fmt.Printf("%s\n", err)
					return
				}
				operator = true
				push(n)
				break
			case 'm':
				operator = true
				mathfunc(s)
				break
			case 'x':
				operator = true
				push(x)
				break
			case '+':
				operator = true
				push(pop() + pop())
				break
			case '*':
				operator = true
				push(pop() * pop())
				break
			case '-':
				operator = true
				op2 := pop()
				push(pop() - op2)
				break
			case '/':
				operator = true
				op2 := pop()
				if op2 != 0.0 {
					push(pop() / op2)
				} else {
					fmt.Printf("error: zero divisor\n")
					break GetOp
				}
				break
			case '%':
				operator = true
				op2 := pop()
				if op2 != 0.0 {
					push(math.Mod(pop(), op2))
				}
				break
			case 'p':
				operator = false
				peek()
				break
			case 'd':
				operator = false
				dup()
				break
			case 's':
				operator = false
				swap()
				break
			case 'c':
				operator = false
				clear()
				break
			case '=':
				pop()
				if v >= 'A' && v <= 'Z' {
					vars[v-'A'] = pop()
				} else {
					fmt.Printf("error: no variable name\n")
					break GetOp
				}
				break
			case '\n':
				if operator {
					x = pop()
					fmt.Printf("\t%.8g\n", x)
				}
				break
			default:
				if typ >= 'A' && typ <= 'Z' {
					push(vars[typ-'A'])
				} else if typ == 'v' {
					push(x)
				} else {
					fmt.Printf("error: unknown command %s\n", string(s))
					break GetOp
				}
			}
			v = typ
		}
	}
}

var j int

func getop(s *[]byte, line []byte) (rune, error) {
	var i int
	var c byte
	if rune(line[j]) == '\n' {
		j++
		return '\n', nil
	}
	for {
		if j >= len(line)-1 {
			break
		}
		c = line[j]
		j++
		if rune(c) != ' ' && rune(c) != '\t' {
			*s = append(*s, c)
			break
		}
	}
	if unicode.IsLower(rune(c)) {
		for {
			c = line[j]
			j++
			i++
			*s = append(*s, c)
			if !unicode.IsLower(rune(c)) {
				break
			}
			if len(string(*s)) > 1 {
				return 'm', nil
			}
			return rune(c), nil
		}
	}
	if c == (*s)[len(*s)-1] {
		j = 0
		return rune(c), nil
	}
	if !unicode.IsDigit(rune(c)) && rune(c) != '.' {
		return rune(c), nil
	}
	if unicode.IsDigit(rune(c)) {
		for {
			c = line[j]
			j++
			i++
			*s = append(*s, c)
			if !unicode.IsDigit(rune(c)) {
				break
			}
		}
	}
	if rune(c) == '.' {
		for {
			c = line[j]
			j++
			i++
			*s = append(*s, c)
			if !unicode.IsDigit(rune(c)) {
				break
			}
		}
	}
	return 'n', nil
}

func mathfunc(s []byte) {
	var op2 float64
	if string(s) == "sin" {
		push(math.Sin(pop()))
	} else if string(s) == "cos" {
		push(math.Cos(pop()))
	} else if string(s) == "exp" {
		push(math.Exp(pop()))
	} else if string(s) == "pow" {
		op2 = pop()
		push(math.Pow(pop(), op2))
	} else {
		fmt.Printf("error: %s is not supported\n", string(s))
	}
}

func peek() {
	x := val[sp-1]
	fmt.Printf("%g\n", x)
}

func dup() {
	push(val[sp])
}

func swap() {
	tmp := val[sp-2]
	val[sp-2] = val[sp-1]
	val[sp-1] = tmp
}

func clear() {
	sp = 0
}

func push(f float64) {
	if sp < MaxVal {
		val[sp] = f
		sp++
	} else {
		fmt.Printf("error: stack full, can't push %f\n", f)
	}
}

func pop() float64 {
	if sp > 0 {
		v := val[sp]
		sp--
		return v
	} else {
		fmt.Printf("error: stack empty\n")
		return 0.0
	}
}

func getLine(lim int) ([]byte, int) {
	var c byte
	var i int
	s := make([]byte, lim)
	r := bufio.NewReader(os.Stdin)
	for i = 0; i < lim-1; i++ {
		c, err := r.ReadByte()
		if rune(c) == '\n' || err != nil {
			break
		}
		s[i] = c
	}
	if rune(c) == '\n' {
		s[i] = c
		i++
	}
	return s, i
}
